### PR TITLE
Caprese reorganization

### PR DESCRIPTION
**Pull request by Robbybp**
## Summary/Motivation:
This is essentially a rewrite of Caprese to make it a more "natural," object-oriented extension of Pyomo. Previously, an `NMPCSim` class managed many interactions between the plant and controller models, and had many unnecessary requirements. The `NMPCSim` class still exists, but the overall class structure is much more "modular." This is in the sense that `nmpc.plant` and `nmpc.controller` are instances of a subclass of `Block` that has added methods for NMPC, and no longer directly interact with each other in the same methods. The benefit, besides (in my opinion) more intuitive syntax, is that these subclasses of `Block` may be instantiated individually to access a useful subset of Caprese's functionality, such as adding piecewise-constant constraints or a least-squares objective function, without setting up an entire NMPC simulation.

Note that this relies on the Pyomo 5.7.3 patch release.

## Changes proposed in this PR:
- `DynamicBlock` and `ControllerBlock` subclasses of `Block`, each of which have methods useful for dynamic simulation and control.
- Classes for storing data produced by a rolling horizon simulation
- An `NmpcVar` subclass of `IndexedVar` that adds attributes (such as `setpoint` and `variance`) necessary for NMPC simulation
- Deleted lots of now-redundant code
- Updated tests and example

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: idaes/apps/caprese/rolling.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        self.time = TimeList(time, tolerance=tolerance)
+
+        super().__init__(data)
+
+    def dim(self):
+        return len(self.keys())
+
+    def __len__(self):
+        # Behavior of len/dim could be confusing. This needs to be
+        # well-documented.
```
Comment (blnicho) [inline idaes/apps/caprese/rolling.py:200]: Could you add docstrings to `__len__` and `dim` to address this comment?

File: idaes/apps/caprese/tests/test_simple_model.py [hunk 2]
```diff
@@ -14,11 +14,14 @@
 Test for Cappresse's module for NMPC.
```
Comment (blnicho) [inline idaes/apps/caprese/tests/test_simple_model.py:14]: Could you fix this old spelling of Caprese?

**Review by blnicho — COMMENTED** [review]
I'm about halfway through this PR. Here are my comments so far.

File: idaes/apps/caprese/dynamic_block.py [hunk 3]
```diff
@@ -0,0 +1,787 @@
...
+                    time.first(),
+                    time.last(),
+                    dae_vars=self.dae_vars,
+                    time_linking_vars=list(self.differential_vars[:]),
+                    outlvl=outlvl,
+                    solver=solver,
+                    )
+
+    def initialize_samples_by_element(self, samples, solver, **kwargs):
+        """ Solve the square problem with fixed inputs for the specified sambles
```
Comment (blnicho) [inline idaes/apps/caprese/dynamic_block.py:383]: Typo: samples

File: idaes/apps/caprese/examples/cstr_nmpc.py [hunk 4]
```diff
@@ -65,116 +70,168 @@ def plot(self):
...
-    
-        S_info = plant._NMPC_NAMESPACE.var_locator[
-                plant.cstr.outlet.conc_mol[0.,'S']]
-        S_location = S_info.location
-        S_group = S_info.group
-        S_data = PlotData(S_group, S_location, name='S_conc')
-        fig, ax = S_data.plot()
-        fig.savefig(S_data.name)
+        nmpc.plant.initialize_by_solving_elements(solver)
+        solver.solve(nmpc.plant)
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/apps/caprese/examples/cstr_nmpc.py:232]: Almost nothing. It's just to reduce any remaining infeasability if any of the subproblems terminated "acceptable" (or infeasible, which is rare but I've seen happen).

File: idaes/apps/caprese/tests/test_dynamic_block.py [hunk 5]
```diff
@@ -0,0 +1,1082 @@
...
+                inputs=inputs_map,
+                measurements=measurements_map,
+                )
+        # Make sure we have the right type
+        assert type(block) is IndexedDynamicBlock
+        assert isinstance(block, DynamicBlock)
+
+        block.construct()
+        # Why is block._data empty here?
+        # Are blocks constructed sparse by default?
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/apps/caprese/tests/test_dynamic_block.py:160]: Yeah I think this is what led to me manually calling block.construct(). Removed these comments

File: idaes/apps/caprese/controller.py [hunk 6]
```diff
@@ -0,0 +1,302 @@
...
+        self.setpoint_objective.activate()
+
+        # Solve single-time point optimization problem
+        dof = degrees_of_freedom(model)
+        if require_steady:
+            assert dof == len(self.INPUT_SET)
+        else:
+            assert dof == (len(self.INPUT_SET) +
+                    len(self.DIFFERENTIAL_SET))
+        results = solver.solve(self, tee=True)
```
[the remark this answers is not in the document]
Comment (Robbybp) [reply idaes/apps/caprese/controller.py]: Added a config option. Still have to decide exactly how I want to do logging in this class and utility functions. In particular, some of my utility functions with embedded solves have non-standard logging, which I need to fix then send the proper options from this class.

File: idaes/apps/caprese/rolling.py (hunk 1 above)
Comment (Robbybp) [reply idaes/apps/caprese/rolling.py]: Done

**Review by blnicho — APPROVED** [review]

**Review by jsiirola — APPROVED** [review]

**Pull request merged**